### PR TITLE
fix float negative zero equality

### LINKED ISSUE #165: 0.0 = -0.0 but the same is not true for floats
I'm not entirely certain if this intentional or not, but I strongly suspect that this is a bug considering it's potential to induce confusion among users.

```
PartiQL> -0.0 = 0.0
   | 
===' 
true
--- 
OK! (0 ms)
PartiQL> `0e0` = `-0e0`
   | 
===' 
false
--- 
OK! (0 ms)
```

### PR DESCRIPTION
**Pull request by raganhan**
Add equality tests for equivalent values and null values

Resolves #165 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by raganhan** [comment]
# [Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=h1) Report
> Merging [#227](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=desc) into [master](https://codecov.io/gh/partiql/partiql-lang-kotlin/commit/774863f73d06f7065166801c62bff4bff46c56e0&el=desc) will **increase** coverage by `0.01%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master     #227      +/-   ##
============================================
+ Coverage     83.32%   83.33%   +0.01%     
- Complexity     1201     1203       +2     
============================================
  Files           152      152              
  Lines          8702     8703       +1     
  Branches       1327     1328       +1     
============================================
+ Hits           7251     7253       +2     
  Misses         1003     1003              
+ Partials        448      447       -1     
```

| Flag | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| #CLI | `53.76% <ø> (ø)` | `44.00 <ø> (ø)` | :arrow_up: |
| #EXAMPLES | `75.34% <ø> (ø)` | `27.00 <ø> (ø)` | :arrow_up: |
| #LANG | `85.09% <100.00%> (+0.01%)` | `975.00 <0.00> (+2.00)` | :arrow_down: |
| #PTS | `100.00% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #TEST_SCRIPT | `79.68% <ø> (ø)` | `157.00 <ø> (ø)` | :arrow_up: |


------

[Continue to review full report at Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=footer). Last update [774863f...537723d](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by raganhan** [comment]
# [Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=h1) Report
> Merging [#227](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=desc) into [master](https://codecov.io/gh/partiql/partiql-lang-kotlin/commit/774863f73d06f7065166801c62bff4bff46c56e0&el=desc) will **increase** coverage by `0.01%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master     #227      +/-   ##
============================================
+ Coverage     83.32%   83.33%   +0.01%     
- Complexity     1201     1203       +2     
============================================
  Files           152      152              
  Lines          8702     8703       +1     
  Branches       1327     1328       +1     
============================================
+ Hits           7251     7253       +2     
  Misses         1003     1003              
+ Partials        448      447       -1     
```

| Flag | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| #CLI | `53.76% <ø> (ø)` | `44.00 <ø> (ø)` | :arrow_up: |
| #EXAMPLES | `75.34% <ø> (ø)` | `27.00 <ø> (ø)` | :arrow_up: |
| #LANG | `85.09% <100.00%> (+0.01%)` | `975.00 <0.00> (+2.00)` | :arrow_down: |
| #PTS | `100.00% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #TEST_SCRIPT | `79.68% <ø> (ø)` | `157.00 <ø> (ø)` | :arrow_up: |


------

[Continue to review full report at Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=footer). Last update [774863f...537723d](https://codecov.io/gh/partiql/partiql-lang-kotlin/pull/227?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 1]
```diff
@@ -180,6 +180,7 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
                     lVal.isPosInf && rVal.isPosInf -> return EQUAL
                     lVal.isPosInf -> return MORE
                     rVal.isPosInf -> return LESS
+                    lVal.isZero() && rVal.isZero() -> 0 // for negative zero
```
Comment (therapon) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt]: As is, the branch does not return and I think the value of `0` is lost (not kept in a variable or returned). 

Also, does the spec state what PartiQL should do for comparing representations of `0` that are in different types, e.g., `0`.compareTo(`-0e0`)? 

If we are to treat all values that denote 0 and -0 as EQUAL then line 183 should `return EQUAL`, if not, I would expect us to take the absolute value (drop the -) and then delegate to the types implementation for comparison.

File: lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt [hunk 2]
```diff
@@ -64,6 +65,7 @@ class NaturalExprValueComparatorsTest : EvaluatorTestBase() {
             "-0.0",
             "`-0.0000000000`",
             "`0e0`",
+            "`-0e0`",
             "`0d10000`",
             "0"
```
Comment (therapon) [inline lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt]: I would add `-0` too here.

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt (hunk 1 above)
Comment (raganhan) [reply lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt]: Fixed. AFAIK PartiQL equality uses the mathematical definition, e.g. `0`, `-0`, `0.000`, `-0.0`, `0e0`, `0.000000000` are all equal

File: lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt (hunk 2 above)
Comment (raganhan) [reply lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt]: Added. `-0` is not an ion value though you can only have a float or decimal negative zero

Comment (therapon) [reply lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt]: My reading of the section on Ion Int has -0 as an example 
http://amzn.github.io/ion-docs/docs/spec.html#int

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 3]
```diff
@@ -180,6 +180,7 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
                     lVal.isPosInf && rVal.isPosInf -> return EQUAL
                     lVal.isPosInf -> return MORE
                     rVal.isPosInf -> return LESS
+                    lVal.isZero() && rVal.isZero() -> return 0 // for negative zero
```
Comment (therapon) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt]: nit: Use `EQUAL` that would keep this function consistent
